### DISCUSSION TITLE
GLSP Theia Client (java-emf-theia): GLSPCommandStack: Failed to execute command: add@http://127.0.0.1:3000/...

### DISCUSSION THREAD

**Discussion by mrcalvin** [eclipse-glsp/glsp#1304] [General]
Hi everyone!

First, let me say that I am impressed by GLSP and your team/ community efforts around it. I enjoy programming in and with it (despite my Java being rusty, my TypeScript basically being non-existent, and my epic personal battle with Maven).

I started working on modifying the [java-emf-theia](https://github.com/eclipse-glsp/glsp-examples/tree/master/project-templates/java-emf-theia), threw out the EMF stuff, and started introducing my own (non-EMF) source model etc. Now I experience the following client-side issue:

<img width="810" alt="image" src="https://github.com/eclipse-glsp/glsp/assets/442057/5538a663-3943-4ea0-a0b7-51409af62a48">

This happens at different points in time (sometimes sooner, sometimes later) when creating new nodes from the palette. The result is that the current diagram freezes, I need to reopen it and start all over (with the state being gone, unsaved etc.)

Did I break the command execution machinery, by introducing my own `GModelOperationHandler` for the `CreateNodeOperation`? (I am flying blind here, as the tutorial here is still in GLSP < 2 state, as the disclaimer states).

```java
public class CreateFeatureNodeHandler extends GModelOperationHandler<CreateNodeOperation>
   implements CreateOperationHandler<CreateNodeOperation> {

   private static Logger LOGGER = LogManager.getLogger(CreateFeatureNodeHandler.class.getSimpleName());

   @Inject
   protected GFModelState modelState;

   protected void createFeature() {
      final IFeatureModelFactory factory = DefaultFeatureModelFactory.getInstance();
      final IFeatureModel fm = modelState.getFModel();
      final String name = "new_one" + String.valueOf(fm.getNumberOfFeatures());
      final IFeature feature = factory.createFeature(fm, name);
      fm.addFeature(feature);
      LOGGER.info("created 2: " + feature.getName());
      /* TODO create the grep also? */
   }

   @Override
   public Optional<Command> createCommand(final CreateNodeOperation operation) {
      return commandOf(() -> createFeature());
   }

   @Override
   public List<String> getHandledElementTypeIds() { return (Arrays.asList(TaskListModelTypes.FEATURE)); }
}
``` 
How can I find the culprit?

Full [log.txt](https://github.com/eclipse-glsp/glsp/files/14863026/log.txt)

**Comment by tortmayr** [comment]
Hi @mrcalvin,
at first glance your `CreateFeatureNodeHandler` implementation looks correct and should work (or at least not throw an exception on execution). Could you please also share the server logs to get more insights? They should be located in `glsp-client/tasklist-theia/logs`

It might be hard to find the cultprit by just relying on logs. I would recommend to debug into the `CreateFeatureNodeHandler` to figure out what is going wrong.

To do this you can start the server manually (in debug mode) in an IDE of your choice (Launch configurations for VS Code and Eclipse are already part of the project template). Then you have to start the Theia application in `external server` mode.
Either via commandline with
```console
yarn start:external
```
or with the `Launch TaskList Theia Backend (External GLSP Server)` launch configuration

>Did I break the command execution machinery, by introducing my own GModelOperationHandler for the CreateNodeOperation? (I am flying blind here, as the tutorial here is still in GLSP < 2 state, as the disclaimer states).

Command execution (and node creation) should still work with your approach. However, the `GModelOperationHandler` is intended for
diagrams that directly use the `GModel` as source model. The `commandOf` method creates a recording command that records changes done on the graphical model. This does not work for your case since you are using a custom source model. No changes will be recorded and therefore undo/redo will also not work.

For now this is ok. But once you have fixed/found the cultprint and node creation is working as expected you probably have to adapt this and use OperationHandlers/Commands that are scoped to your source model.

First I would recommend to create abstract base classes for GF-based operation handlers:

```java
public abstract class GFOperationHandler<O extends Operation> extends BasicOperationHandler<O> {

   // Put common functionality e.g. utility methods or props that will be used in (most of) the subclasses
   @Inject()
   protected GFModelState modelState;

   @Inject()
   protected ActionDispatcher actionDispatcher;
}
```

```java
public abstract class GFCreateOperationHandler<T extends CreateOperation>
   extends GFOperationHandler<T> implements CreateOperationHandler<T> {

   protected List<String> handledElementTypeIds;

   public GFCreateOperationHandler(final String... elementTypeIds) {
      this(Lists.newArrayList(elementTypeIds));
   }

   public GFCreateOperationHandler(final List<String> handledElementTypeIds) {
      this.handledElementTypeIds = handledElementTypeIds;
   }

   @Override
   public List<String> getHandledElementTypeIds() { return handledElementTypeIds; }

   public void setHandledElementTypeIds(final List<String> handledElementTypeIds) {
      this.handledElementTypeIds = handledElementTypeIds;
   }

   // Additional common utilities for create handlers

}
```
Then you can derive the `CreateFeatureNodeHandler` from this base class and implement a custom command:
```java
...
import org.eclipse.emf.common.command.AbstractCommand;

public class CreateFeatureNodeHandler extends GFCreateOperationHandler<CreateNodeOperation> {

   public CreateFeatureNodeHandler() {
      super(TaskListModelTypes.FEATURE);
   }

   @Override
   public Optional<Command> createCommand(final CreateNodeOperation operation) {
     return new CreateFeatureCommand(this.modelState.getFModel());
   }

   class CreateFeatureCommand extends AbstractCommand {
      private final IFeatureModel model;
      private IFeature addedFeature;

      CreateFeatureCommand(final IFeatureModel model) {
         this.model = model;
      }

      @Override
      public void execute() {
         final IFeatureModelFactory factory = DefaultFeatureModelFactory.getInstance();
         final String name = "new_one" + String.valueOf(model.getNumberOfFeatures());
         final IFeature feature = factory.createFeature(model, name);
         model.addFeature(feature);
         this.addedFeature = feature;
         LOGGER.info("created 2: " + feature.getName());

      }

      @Override
      public void redo() {
         if (this.addedFeature != null) {
            model.addFeature(feature)
         }

      }

      @Override
      public void undo() {
         if (this.addedFeature != null) {
            model.removeFeature(addedFeature)
         }
      }

   }

}
```
(Command implementation is just a rough sketch and probably needs some adaption to work properly)

**Reply by mrcalvin** [reply]
Hi @tortmayr 

Thank you for the extensive, helpful answer!

> But once you have fixed/found the cultprint and node creation is working as expected you probably have to adapt this and use OperationHandlers/Commands that are scoped to your source model.

I will update my handler implementation accordingly. (Started wondering already why undo/ redo did not work.) A few clarifications:

> I would recommend to debug into the CreateFeatureNodeHandler to figure out what is going wrong.

I see. But there are no exceptions thrown in the server-side handler itself, but on the client-side command stack equivalent, right? So, how would I debug into this one? What is a good starting point (dispatcher)?

> Could you please also share the server logs to get more insights? They should be located in glsp-client/tasklist-theia/logs

I'd love to, but the directory is empty, no log file is being produced? In VS, when runng the debug target `Launch TaskList Theia Backend (External GLSP Server)`, I see the "Debug Console" being populated. No log file. `yarn server:external` fails for me with:

```
2024-04-15T14:38:43.748Z root ERROR Uncaught Exception:  Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
2024-04-15T14:38:43.748Z root ERROR Error [ERR_STREAM_DESTROYED]: Cannot call write after a stream was destroyed
    at new NodeError (node:internal/errors:405:5)
    at _write (node:internal/streams/writable:324:11)
    at Writable.write (node:internal/streams/writable:337:10)
    at /Users/ssoberni/Documents/dev/fm4/client/tasklist-browser-app/lib/backend/vendors-node_modules_theia_core_lib_common_index_js.js:15382:29
    at new Promise (<anonymous>)
    at WritableStreamWrapper.write (/Users/ssoberni/Documents/dev/fm4/client/tasklist-browser-app/lib/backend/vendors-node_modules_theia_core_lib_common_index_js.js:15372:16)
    at SocketMessageWriter.doWrite (/Users/ssoberni/Documents/dev/fm4/client/tasklist-browser-app/lib/backend/vendors-node_modules_theia_core_lib_common_index_js.js:14482:33)
    at /Users/ssoberni/Documents/dev/fm4/client/tasklist-browser-app/lib/backend/vendors-node_modules_theia_core_lib_common_index_js.js:14473:29
2024-04-15T14:38:43.752Z root ERROR [JsonrpcGLSPClient] Connection to server got closed. Server will not be restarted. []
```
🤔
